### ISSUE TITLE
Make sure that underlying goroutine finishes when requests times out

### ISSUE THREAD

**Issue by marseel** [kubernetes/kubernetes#105884]
### What happened?

Spin off from #105611 as requested in https://github.com/kubernetes/kubernetes/pull/105734#discussion_r732852756

Currently, timeout filter doesn't force underlying goroutine to finish. This can potentially cause hard to debug race conditions as showed in #105734 

### What did you expect to happen?

Timeout goroutine should wait for underlying goroutine to finish first.

### How can we reproduce it (as minimally and precisely as possible)?

N/A

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by wojtek-t** [comment]
>  Personally, I think it's acceptable, because clients handle timeouts by themselves too and we could also fix cases when goroutine A is not honoring timeout.

It depends - if we run it a bit more (say hundreds milliseconds) - that is fine.
If it keeps running for another 3 minutes - then it's not fine...

**Comment by mborsz** [comment]
https://github.com/golang/go/issues/16100#issuecomment-255498503 contains some suggested solution for the issue of blocking Write/Read:
> Do all your I/O in separate goroutines, and panic if they take too long?
 

If blocking transformResponseObject is the only problem now, this may be a reasonable solution.

**Comment by mborsz** [comment]
/remove-lifecycle rotten

> yes, we have seen the child go routine executing the rest of the handler chain to be actively doing stuff even after the request had been timed out by the `timeout` filter. one example is `transformResponseObject` - we see the writing of response for a big list goes on even after the request times out.
> 
> https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go#L285-L286
> 
> since `http.ResponseWriter.Write` is not context enabled yet. On the other hand, i am not aware of any mechanism with which one can kill a goroutine from the outside.

We discussed this problem with golang team recently and they prepared a proposal to make it possible to manipulate per-request timeouts (so that we can set WriteDeadline on per-request basis) to avoid Write() call live longer than the request deadline: https://github.com/golang/go/issues/54136

@tkashem IIUC setting WriteDeadline for non-long-running requests to `time.Now() + request-timeout` should solve issue of goroutines getting stuck in Write() call.

Could you help me verifying if this is true?

How about long-running requests (such as watches)? For them we can't set WriteDeadline as they may run potentially for a long time, but also we don't set request timeout for them, right? So maybe having this proposal implemented is sufficient to get rid of the timeoutHandler?

**Comment by tkashem** [comment]
> IIUC setting WriteDeadline for non-long-running requests to time.Now() + request-timeout should solve issue of goroutines getting stuck in Write() call.

@mborsz if I am not mistaken, `WriteDeadline` is a `Server` level property, and is used to set the write deadline at a connection level inside the net/http `readRequest` method
https://github.com/golang/go/blob/e1fd51e0765a5d7b2dd72b2f82b9c3eed2764035/src/net/http/server.go#L969-L974

so as you have metioned above, today we can not set `WriteDeadline` on per-request basis. But with the https://github.com/golang/go/issues/54136 implemented, I believe we will be able to set `WriteDeadline` for each request 
```
RequestReceivedTimestamp + {timeout specified in request URI | request-timeout}
```

> How about long-running requests (such as watches)? For them we can't set WriteDeadline as they may run potentially for a long time, but also we don't set request timeout for them, right?

I think we can:
- set `WriteDeadline` for a long running request that has a `timeout` in the request URI (per-request basis)
- technically, we can also set `WriteDeadline` for individual watch requests

**Comment by lavalamp** [comment]
> Do all your I/O in separate goroutines, and panic if they take too long?

I'd expect closing the connection externally would cause anything trying to read/write to it to fail and thus be unblocked?

**Comment by tkashem** [comment]
> @tkashem IIUC setting WriteDeadline for non-long-running requests to time.Now() + request-timeout should solve issue of goroutines getting stuck in Write() call.
>
> Could you help me verifying if this is true?

@mborsz when Go 1.10 adds the new `ResponseController` then we can wire it up: https://github.com/kubernetes/kubernetes/pull/114189

**Comment by wojtek-t** [comment]
/remove-lifecycle rotten

@tkashem @mborsz - you meant golang 1.20? [go 1.10 is a historical version]...

**Comment by wojtek-t** [comment]
FWIW - yes, it will be 1.20:
https://github.com/golang/go/issues/54136#issuecomment-1320339696

**Comment by tkashem** [comment]
@wojtek-t it was a typo, thanks for pointing it out.

**Comment by tkashem** [comment]
Note: in addition to the timeout filter, we have `finisher` which spins up go routine to execute the request logic, one example:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go#L188

Once we wire `ResponseController`, i think we might be able to remove it.

**Comment by BenTheElder** [comment]
/triage accepted
@wojtek-t did we ever follow up on https://github.com/kubernetes/kubernetes/issues/105884#issuecomment-1370922570 ?

**Comment by wojtek-t** [comment]
Yes - there is an approved KEP for it: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4460-per-request-deadline

But I don't think the implementation landed...

@tkashem - FYI

**Comment by aaron-prindle** [comment]
/triage accepted

@wojtek-t - any updates on the KEP implementation for the issue here?
